### PR TITLE
Functionalize mutations in deferred local_map

### LINKED ISSUE #167568: Autoparallel's local_map doesn't functionalize triton kernels properly
### 🐛 Describe the bug

Note this only affects Autoparallel aka _DEFER_INLINING option

This works for cond:

<details><summary>Details</summary>
<p>

```python
def test_mutations_triton_hop(self):
        import torch
        import triton
        import triton.language as tl

        @triton.jit
        def add_scalar_inplace_kernel(Y_ptr, addend, n_elems, BLOCK: tl.constexpr):
            pid = tl.program_id(0)
            offs = pid * BLOCK + tl.arange(0, BLOCK)
            mask = offs < n_elems
            y = tl.load(Y_ptr + offs, mask=mask)
            y = y + addend
            tl.store(Y_ptr + offs, y, mask=mask)

        def triton_add_inplace_(y, addend=10000):
            n = y.numel()
            # addend_t = y.new_tensor(addend)
            BLOCK = 1024
            grid = lambda meta: (triton.cdiv(n, meta["BLOCK"]),)
            add_scalar_inplace_kernel[grid](y, addend, n, BLOCK=BLOCK)
            return y

        def true_fn(x):
            out = x + 10
            triton_add_inplace_(out)
            return out

        def false_fn(x):
            out = x + 100
            triton_add_inplace_(out)
            return out

        def fn(pred, pytree_in):
            return torch.cond(
                pred, true_fn, false_fn, (pytree_in,)
            )

        x = torch.randn(10, 100, device="cuda", requires_grad=True)
        torch.compile(fn)(torch.tensor(True), x)
```

</p>
</details> 

But not for local_map:

<details><summary>Details</summary>
<p>

```python
def test_mutations_triton_hop(self):
        import torch
        import triton
        import triton.language as tl

        @triton.jit
        def add_scalar_inplace_kernel(Y_ptr, addend, n_elems, BLOCK: tl.constexpr):
            pid = tl.program_id(0)
            offs = pid * BLOCK + tl.arange(0, BLOCK)
            mask = offs < n_elems
            y = tl.load(Y_ptr + offs, mask=mask)
            y = y + addend
            tl.store(Y_ptr + offs, y, mask=mask)

        def triton_add_inplace_(y, addend=10000):
            n = y.numel()
            # addend_t = y.new_tensor(addend)
            BLOCK = 1024
            grid = lambda meta: (triton.cdiv(n, meta["BLOCK"]),)
            add_scalar_inplace_kernel[grid](y, addend, n, BLOCK=BLOCK)
            return y

        @local_map(
            out_placements=((Replicate(), Replicate(), Replicate()),),
            in_placements=(
                (Replicate(), Replicate(), Replicate()),
                (Replicate(), Replicate(), Replicate()),
            ),
            redistribute_inputs=True,
            in_grad_placements=None,
            device_mesh=self.mesh,
        )
        def replicate_linear(w, x):
            y = torch.matmul(x, w.t())
            triton_add_inplace_(y)
            return y

        class MyModule(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.w = nn.Linear(80, 80)

            def forward(self, x):
                return replicate_linear(self.w.weight, x)

        model = MyModule()

        def inputs_fn():
            return (torch.randn(80, 80),)

        joint = ap_style_initial_capture(model, inputs_fn)
        print(joint)
        local_map_nodes = [n for n in joint.graph.nodes if "local_map" in n.name]
        fw_subgraph = local_map_nodes[0].meta["debug_gm"].graph
        bw_subgraph = local_map_nodes[1].meta["debug_gm"].graph
```

</p>
</details> 

Because we're tracing the joint graph inside the HOP and disabling functionalization, the kernel will be desugared away

### Versions

main

cc @chauhang @penguinwu @ydwu4 @bdhirsh

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185043

Deferred local_map builds its joint forward/backward graph from an already captured local body while functionalization is suspended. Mutation nodes can therefore survive into partitioning, and side-effect-only Triton mutation HOPs can be dropped entirely under FakeTensor execution. A raw joint trace can also fail when an in-place update changes a tensor that an earlier operator saved for backward.

Materialize a forward-only graph first so call_method, mutable ATen, out=, Python wrapper, and mutation-HOP forms are visible. For supported non-invoke bodies, functionalize that forward before AD tracing, using an explicit empty make_fx decomposition table so detach boundaries survive into the later backward. Bodies containing opaque Python callables stay on the AD-aware path so a hidden custom autograd function keeps its backward semantics. Only the exact torch package namespace is treated as internal, so third-party packages such as torchao do not accidentally take the pre-AD path.

invoke_subgraph needs Python-mode functionalization but saves operands for lazy backward construction. Give inspection and joint traces fresh cache identifiers, and use stride-preserving operand clones only while building its mutation-bearing joint. After custom autograd and nested HOP backward logic has expanded, functionalize the canonical joint before partitioning. Effectful or still-unsupported HOP combinations fail explicitly instead of silently miscompiling.

Pre-functionalizing every body was rejected because it can erase custom autograd boundaries, while post-joint-only functionalization can hit saved-tensor version checks before it gets a graph. The split path preserves AD semantics while normalizing mutations early whenever that is safe.

Fixes #167568
Generated by my agent

Test Plan:
- `python test/higher_order_ops/test_local_map.py TestLocalMap.test_mutation_with_opaque_custom_autograd`
- `python test/higher_order_ops/test_local_map.py`
- `source .venv/bin/activate && lintrunner -a`
- `git diff --check`

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by xmfan — APPROVED** [review]

**Comment by jansel** [comment]
@zou3519 can you look at this one?

**Comment by zou3519** [comment]
yup will do tomorrow

File: torch/_higher_order_ops/local_map.py [hunk 1]
```diff
@@ -242,6 +242,69 @@ def create_hop_fw_bw(
...
+    )
+
+    def run_fw_gm(*args: Any) -> Any:
+        # The deferred local_map path builds a joint graph from the already
+        # captured local body. Functionalize Triton mutation wrappers while
+        # tracing the joint so they are preserved as functional HOPs instead of
+        # disappearing under FakeTensorMode.
+        fw_callable: Callable[..., Any] = torch.fx.Interpreter(fw_gm).run
+        if needs_triton_functionalization:
+            fw_callable = torch.func.functionalize(fw_callable, remove="mutations")
```
Comment (zou3519) [inline torch/_higher_order_ops/local_map.py]: @xmfan This PR only applies functionalization (on the entire forward) if triton kernels are in the forward pass. Otherwise, it does not apply functionalization.

That doesn't seem right to me as a fix? I don't think it's good that the presence of triton kernels leads to functionalization. Ideally we always use functionalization, or we always don't use functionalization, and we figure something else out.

That being said, this is local_map, so I will defer to you if you want this behavior

**Review by zou3519 — CHANGES_REQUESTED** [review]
requesting review to have a discussion

File: torch/_higher_order_ops/local_map.py (hunk 1 above)
Comment (zou3519) [reply torch/_higher_order_ops/local_map.py]: for example, mutable ops seem like they can also run into this problem, so at the very least we should agree on what to do in that case too. https://github.com/pytorch/pytorch/issues/182403

Comment (zou3519) [reply torch/_higher_order_ops/local_map.py]: I have also discovered https://github.com/pytorch/pytorch/pull/184409, which always functionalizes

Comment (jansel) [reply torch/_higher_order_ops/local_map.py]: My agent says I changed the fix to key off captured mutations rather than Triton specifically: deferred local_map now functionalizes bodies with mutable ATen ops or functionalizable mutation HOPs, while leaving non-mutating bodies on the existing path. I also added an index_put_ regression for the ordinary mutable-op case you called out.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the two docs failures completed docs generation and failed during preview upload with `aws: command not found`, which is unrelated to this local_map change. I reran the failed docs jobs and will wait for that rerun before re-requesting review.

File: torch/_higher_order_ops/local_map.py (hunk 1 above)
Comment (xmfan) [reply torch/_higher_order_ops/local_map.py]: We should functionalize all the time, but you need to verify autoparallel still works, and it's something I can do after this PR lands @zou3519

Comment (morrison-turnansky) [reply torch/_higher_order_ops/local_map.py]: @Arindamsamanta2004 going to post here to consolidate the discussion from https://github.com/pytorch/pytorch/pull/184409.

tldr My take is that if this PR doesn't break autoparallel it is the better approach. 

The PRs are similar and should both fix the mutation problem, but they suffer from different limitations. 

The 2 pass approach of https://github.com/pytorch/pytorch/pull/184409 major problem is that it can lose data. Maybe this is a fixable problem, but it could be hacky. It will also have more overhead.  However since it does the functionalization outside the trace, it shouldn't break autoparallel. 

This pr with the one pass approach is able to avoid the metadata loss issue because it functionalizes during the trace as opposed to a 2 stage. However this runs into the problem of effecting autoparallel 

If this PR doesn't break invariants of local map (autoparallel), then it is the more rigorous approach and should cover error in https://github.com/pytorch/pytorch/issues/182403. Candidly, I am not super familiar with design decisions of local map, but if we can functionalize during trace and support autoparallel that is the more maintainable/elegant approach.

Comment (jansel) [reply torch/_higher_order_ops/local_map.py]: My agent says the current patch is deliberately mutation-gated: deferred local_map uses the existing path for non-mutating bodies, and only runs functionalization when the captured body contains mutable ATen ops or functionalizable mutation HOPs. The added regressions exercise the AP-style deferred capture path for both the Triton mutation from this issue and an ordinary `index_put_` mutation, and the PR CI has passed the relevant test jobs; the remaining failures are docs preview upload steps failing with `aws: command not found` after docs generation.

**Comment by jansel** [comment]
My agent says the rerun of the two docs jobs failed in the same upload-only step: both jobs completed docs generation and then hit `aws: command not found` while syncing previews to S3. The PR changes are limited to `local_map.py` and `test_local_map.py`, and all non-docs checks shown by `gh pr checks` are passing or skipped.

**Review by sanketpurandare — APPROVED** [review]

File: torch/_higher_order_ops/local_map.py (hunk 1 above)
Comment (jansel) [reply torch/_higher_order_ops/local_map.py]: My agent says the rebased patch now handles Dynamo’s `operator.setitem` capture form in addition to mutable ATen ops and functionalizable mutation HOPs. The updated index-put regression inlines the local_map bodies and fails without this fix; both mutation regressions, the full local_map suite, and lintrunner pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says 
> Addressed the automated `call_method` finding and broadened the fix around the underlying deferred-functionalization boundary. Supported mutation-bearing bodies are normalized before AD when safe; custom-autograd and nested-HOP cases are functionalized after their AD-aware joint is materialized. The update also preserves `detach` in the later backward and protects `invoke_subgraph` cache/version semantics with stride-preserving clones. Added numerical forward/backward regressions for Python mutation forms, `out=`, `cond`, opaque custom autograd, Triton, and nested regions. `python test/higher_order_ops/test_local_map.py` (28 tests), `lintrunner -a`, and `git diff --check` pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says
> Rebased onto current main and verified a genuine issue from the latest automated review: the old `startswith("torch")` check treated third-party namespaces such as `torchao` as PyTorch internals, sending hidden custom-autograd calls through pre-AD functionalization. The check now recognizes only `torch` and `torch.*`; the regression covers `torchao.fake` and verifies custom backward gradients. Direct local_map input mutations are already rejected by Dynamo before this path. The full 28-test local_map suite, `lintrunner -a`, and `git diff --check` pass.
